cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -705,80 +890,39 @@ func StartControllers(ctx context.Context, controllerCtx ControllerContext, cont
-			// HealthChecker should be present when controller has started
-			controllerChecks = append(controllerChecks, check)
+		if err := buildController(controllerDesc); err != nil {
+			return nil, err
 		}
 	}
 
-	healthzHandler.AddHealthChecker(controllerChecks...)
-
-	return nil
+	// Register the checks.
+	if len(checks) > 0 {
+		healthzHandler.AddHealthChecker(checks...)
+	}
+	return controllers, nil
 }
 
-// StartController starts a controller with a specified ControllerContext
-// and performs required pre- and post- checks/actions
-func StartController(ctx context.Context, controllerCtx ControllerContext, controllerDescriptor *ControllerDescriptor,
-	unsecuredMux *mux.PathRecorderMux) (healthz.HealthChecker, error) {
+// StartControllers starts all controllers concurrently and blocks until the context is cancelled and all controllers are terminated.
+func StartControllers(ctx context.Context, controllerCtx ControllerContext, controllers []Controller) {
 	logger := klog.FromContext(ctx)
-	controllerName := controllerDescriptor.Name()
-
-	for _, featureGate := range controllerDescriptor.GetRequiredFeatureGates() {
-		if !utilfeature.DefaultFeatureGate.Enabled(featureGate) {
-			logger.Info("Controller is disabled by a feature gate", "controller", controllerName, "requiredFeatureGates", controllerDescriptor.GetRequiredFeatureGates())
-			return nil, nil
-		}
-	}
-
-	if controllerDescriptor.IsCloudProviderController() {
-		logger.Info("Skipping a cloud provider controller", "controller", controllerName)
-		return nil, nil
-	}
-
-	if !controllerCtx.IsControllerEnabled(controllerDescriptor) {
-		logger.Info("Warning: controller is disabled", "controller", controllerName)
-		return nil, nil
-	}
-
-	time.Sleep(wait.Jitter(controllerCtx.ComponentConfig.Generic.ControllerStartInterval.Duration, ControllerStartJitter))
-
-	logger.V(1).Info("Starting controller", "controller", controllerName)
-
-	initFunc := controllerDescriptor.GetInitFunc()
-	ctrl, started, err := initFunc(klog.NewContext(ctx, klog.LoggerWithName(logger, controllerName)), controllerCtx, controllerName)
-	if err != nil {
-		logger.Error(err, "Error starting controller", "controller", controllerName)
-		return nil, err
-	}
-	if !started {
-		logger.Info("Warning: skipping controller", "controller", controllerName)
-		return nil, nil
-	}
-
-	check := controllerhealthz.NamedPingChecker(controllerName)
-	if ctrl != nil {
-		// check if the controller supports and requests a debugHandler
-		// and it needs the unsecuredMux to mount the handler onto.
-		if debuggable, ok := ctrl.(controller.Debuggable); ok && unsecuredMux != nil {
-			if debugHandler := debuggable.DebuggingHandler(); debugHandler != nil {
-				basePath := "/debug/controllers/" + controllerName
-				unsecuredMux.UnlistedHandle(basePath, http.StripPrefix(basePath, debugHandler))
-				unsecuredMux.UnlistedHandlePrefix(basePath+"/", http.StripPrefix(basePath, debugHandler))
+	var wg sync.WaitGroup
+	wg.Add(len(controllers))
```

Comment:
Wondering if this is the best construct for tracking goroutines spawned by the controllers. While it works for now, it makes it so that we can get stuck if any thread refuses to terminate without any info on which one refuses to terminate. We can solve the termination issue by either doing a hard shutdown if that ever happens but that still gives us no info about the underlying error. 

Not something that needs to be solved immediately but curious whether we can make this better(channels instead of waitgroups maybe?) so that we can make this more debuggable in cases of errors. This is currently the case so we aren't making anything worse but issues will be a lot more apparent with graceful termination so it's in our best interest to expose as much info here.

Reply:
Well, if you ever get stuck, you can only call os.Exit, because in that case you cannot be sure whether you can give up leading without leaving some logic to run in parallel. So I would say this needs to be addressed once we get to aligning that part.

In any case, added some extra logging in cc9eec42d4d , see if that's enough for now.

Reply:
My original idea was to leave this to Kubernetes in case the pod gets stuck, for now. It will get killed after the graceful termination period. And once we get to actually improving leader election code, we would align this to exit when `Run()` gets stuck. But it can be improved now as well. We would just need to come up with a termination period that would be either hard-coded or configurable via a flag. Then the question is what period to choose. Controllers should really exit rather quickly, I guess, but I haven't checked whether they do any sophisticated cleanup. I guess they don't as the current state of matters is to just kill the executable without waiting for anything 🙂

Reply:
@Jefftree Added hard shutdown after a certain timeout, see 43630e4eebb9585113c1ae5a47f5752bc26e50fc

Just hard-coded 10 seconds there for now, open to changing that.

Reply:
@Jefftree Added hard shutdown after a certain timeout, see 23aef544c73d003a98dc2752ba34718de3f8f24a

Just hard-coded 10 seconds there for now, open to changing that.

Sorry for the previous comment, was a broken commit not passing tests.